firstvoices/backend/views/dictionary_views.py

```diff
@@ -52,3 +57,116 @@ def get_queryset(self):
             )
         else:
             return DictionaryEntry.objects.none()
+
+
+@extend_schema_view(
+    list=extend_schema(
+        description="Returns a word of the day for the given site.",
+        responses={
+            200: DictionaryEntryDetailSerializer,
+            403: OpenApiResponse(description="Todo: Not authorized for this Site"),
+            404: OpenApiResponse(description="Todo: Site not found"),
+        },
+    ),
+)
+class WordOfTheDayView(
+    FVPermissionViewSetMixin,
+    SiteContentViewSetMixin,
+    mixins.ListModelMixin,
+    viewsets.GenericViewSet,
+):
+    """
+    Picks and returns word-of-the-day for a given site.
+    """
+
+    http_method_names = ["get"]
+    serializer_class = DictionaryEntryDetailSerializer
+    pagination_class = None
+
+    @staticmethod
+    def get_suitable_queryset(site_slug):
+        # Case 1. Check if there is a word assigned word-of-the-day date of today
+        today = datetime.today()
+        selected_word = WordOfTheDay.objects.filter(
+            site__slug=site_slug, date=today
+        ).first()
+        if selected_word is not None:
+            queryset = DictionaryEntry.objects.filter(
+                id=selected_word.dictionary_entry.id
+            )
+            if queryset.count() > 0:
+                return queryset
+
+        # Case 2. If no words found with today's date, Get words which have not yet been assigned word-of-the-day
+        words_used = WordOfTheDay.objects.filter(site__slug=site_slug).values_list(
+            "dictionary_entry_id", flat=True
+        )
+        queryset = DictionaryEntry.objects.filter(
+            site__slug=site_slug,
+            type=DictionaryEntry.TypeOfDictionaryEntry.WORD,
+            exclude_from_wotd=False,
+        ).exclude(id__in=list(words_used))
+        if queryset.count() > 0:
+            selected_word = queryset.first()
+            WordOfTheDay(
+                date=today, dictionary_entry=selected_word, site=selected_word.site
+            ).save()
+            return queryset
+
+        # Case 3. If no words found satisfying any of the above condition, try to find wotd which has not
+        # been assigned a date in the last year
+        last_year_date = today - timedelta(weeks=52)
+        selected_word = (
+            WordOfTheDay.objects.filter(site__slug=site_slug)
+            .order_by("-date")
+            .distinct()
+            .filter(date__lte=last_year_date)
+            .last()
+        )
+        if selected_word is not None:
+            queryset = DictionaryEntry.objects.filter(
+                id=selected_word.dictionary_entry.id
+            )
+            if queryset.count() > 0:
+                WordOfTheDay(
+                    date=today,
+                    dictionary_entry=selected_word.dictionary_entry,
+                    site=selected_word.dictionary_entry.site,
+                ).save()
```

Comment:
Would this create an ever-growing list of WordOfTheDay models in the db? For example if "Hello" is added as a word of the day today, then a year from now there is another model on the site with "Hello" and a different date, etc.

Reply:
Yes, the goal is a log of them :)